docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,332 @@
+
+Write less E2E tests. End-to-end tests are slow by nature, their amount should be drastically lower than that of unit or integration tests.
+
+## Smart Assertions
+
+In end-to-end web testing, unpredictable factors (like network lag, processor, or memory bottlenecks in containers) can interfere with the assertions and produce inconsistent test results. Such tests are inconclusive (sometimes called 'flaky').
+
+TestCafe includes a [Smart Assertion Query Mechanism](../../guides/basic-guides/assert.md#smart-assertion-query-mechanism). This mechanism introduces extra wait time for all the assertions: if an assertion fails, it retries multiple times within a timeout. That reduces random impact and stabilizes the tests without performance trade-offs.
+
+The following example demonstrates a common mistake:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    await t.typeText(developerNameInput, 'Peter');
+
+    //an awaited selector doesn't update and produces unstable test results. Avoid it.
+    const developerName = await Selector('#developer-name').value;
+    //
+
+    await t
+            .expect(developerName).eql('Peter')
+            .typeText(developerNameInput, 'Jack')
+            .expect(developerName).eql('Jack'); // fails
+});
+```
+
+In this snippet, the `developerName` is initialized with the value of a Selector, but because of the `await` keyword, the value is calculated once and doesn't update. This disables the smart assertions query mechanism, which in this case leads to inconclusive results.
+
+To avoid this, omit the `await` keyword:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    await t.typeText(developerNameInput, 'Peter');
+
+    const developerName = Selector('#developer-name').value;
+
+    await t
+            .expect(developerName).eql('Peter')
+            .typeText(developerNameInput, 'Jack')
+            .expect(developerName).eql('Jack'); // passes
+});
+```
+
+The `developerName` is now initialized with a re-executable Selector API promise. When a test controller receives this promise, it enables the smart assertion query mechanism and TestCafe can wait for the value to update.
+
+> The smart assertion query mechanism works with Client Functions and promises generated by TestCafe Selector API. Use the `await` keyword with user-created promises and promises generated by third-party libraries.
```

Comment:
I wouldn't say that promises are "generated". Data sets can be generated, tokens can be generated (as they are pieces of data), but promises are returned, created, obtained from...

Reply:
> The smart assertion query mechanism works with Client Functions and promises created by TestCafe Selector API. Use the `await` keyword with user-created promises and promises returned from third-party libraries.